### PR TITLE
Fix loading symbolic expressions containing symbolic functions

### LINKED ISSUE #40292: Loading expressions containing symbolic functions is broken
### Steps To Reproduce

```
$ sage
┌────────────────────────────────────────────────────────────────────┐
│ SageMath version 10.7.beta6, Release Date: 2025-06-14              │
│ Using Python 3.12.10. Type "help()" for help.                      │
└────────────────────────────────────────────────────────────────────┘
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Warning: this is a prerelease version, and it may be unstable.     ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
sage: var('x,y')
(x, y)
sage: u = function('u')(x,y)
sage: save(u, 'u')
sage: load('u')
```

### Expected Behavior

`u(x,y)`

### Actual Behavior

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
File ~/src/sage/src/sage/symbolic/function_factory.py:110, in function_factory(name, nargs, latex_name, conversions, evalf_params_first, eval_func, evalf_func, conjugate_func, real_part_func, imag_part_func, derivative_func, tderivative_func, power_func, series_func, print_func, print_latex_func)
    107             raise ValueError(func_name + "_func" + " parameter must be callable")
    108         setattr(NewSymbolicFunction, '_%s_' % func_name, func)
--> 110 return NewSymbolicFunction()

File ~/src/sage/src/sage/symbolic/function_factory.py:55, in function_factory.<locals>.NewSymbolicFunction.__init__(self)
     46 def __init__(self):
     47     """
     48     EXAMPLES::
     49 
   (...)     53         f(2, 4)
     54     """
---> 55     SymbolicFunction.__init__(self, name, nargs, latex_name,
     56             conversions, evalf_params_first)

File ~/src/sage/src/sage/symbolic/function.pyx:1217, in sage.symbolic.function.SymbolicFunction.__init__()
   1215 """
   1216 self.__hinit = False
-> 1217 Function.__init__(self, name, nargs, latex_name, conversions,
   1218         evalf_params_first)
   1219 

File ~/src/sage/src/sage/symbolic/function.pyx:218, in sage.symbolic.function.Function.__init__()
    216     ValueError: eval_func parameter must be callable
    217 """
--> 218 self._name = name
    219 self._alt_name = alt_name
    220 self._nargs = nargs

TypeError: Expected unicode, got bytes
Exception ignored in: 'sage.symbolic.expression.py_get_serial_for_new_sfunction'
Traceback (most recent call last):
  File "/home/rburing/src/sage/src/sage/symbolic/function_factory.py", line 110, in function_factory
    return NewSymbolicFunction()
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/rburing/src/sage/src/sage/symbolic/function_factory.py", line 55, in __init__
    SymbolicFunction.__init__(self, name, nargs, latex_name,
  File "sage/symbolic/function.pyx", line 1217, in sage.symbolic.function.SymbolicFunction.__init__
  File "sage/symbolic/function.pyx", line 218, in sage.symbolic.function.Function.__init__
TypeError: Expected unicode, got bytes
conjugate(x)
```

### Additional Information

_No response_

### Environment

- **OS**: Ubuntu 24.04.2 LTS
- **Sage Version**: 10.7.beta6 (using conda for dependencies)

### Checklist

- [x] I have searched the existing issues for a bug report that matches the one I want to file, without success.
- [x] I have read the documentation and troubleshoot guide

### PR DESCRIPTION
**Pull request by rburing**
- Fixes https://github.com/sagemath/sage/issues/40292
- Fixes https://github.com/sagemath/sage/issues/30018

### :memo: Checklist

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation and checked the documentation preview.

### CODE REVIEW

File: src/sage/misc/persist.pyx [hunk 1]
```diff
@@ -159,6 +159,19 @@ def load(*filename, compress=True, verbose=True, **kwargs):
...
+    Check that :issue:`40292` is fixed::
+
+        sage: # needs sage.symbolic
+        sage: var('x,y')
+        (x, y)
+        sage: u = function('u')(x, y)
+        sage: t = tmp_filename(ext='.sobj')
+        sage: save(u, t)
+        sage: load(t)
+        u(x, y)
```
Comment (user202729) [inline src/sage/misc/persist.pyx]: does testing with dumps/loads work? Avoid the temp file

File: src/sage/symbolic/pynac_impl.pxi [hunk 2]
```diff
@@ -817,7 +817,9 @@ cdef unsigned py_get_serial_for_new_sfunction(stdstring &s,
     create one and set up the function tables properly.
     """
     from sage.symbolic.function_factory import function_factory
-    cdef Function fn = function_factory(s.c_str(), nargs)
+    cdef const char* c_name = s.c_str()
+    py_name = c_name.decode('UTF-8')
```
Comment (user202729) [inline src/sage/symbolic/pynac_impl.pxi]: either add comment to explain where the corresponding `encode()` is (to justify using `UTF-8` here), or add a test with some Unicode characters in function name

Comment (rburing) [reply src/sage/symbolic/pynac_impl.pxi]: It was true that the string was encoded using `UTF-8`, but I changed the code to follow the existing style and I added a comment explaining where the string is encoded.
